### PR TITLE
gzip filter: allow setting zlib compressor's chunk size

### PR DESCRIPTION
**Pull request by rgs1**
Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
fwiw, nginx sets this value to 8kb. see:

http://nginx.org/en/docs/http/ngx_http_gzip_module.html#gzip_buffers

File: source/extensions/filters/http/gzip/gzip_filter.cc [hunk 1]
```diff
@@ -28,10 +28,11 @@ GzipFilterConfig::GzipFilterConfig(const envoy::extensions::filters::http::gzip:
       compression_level_(compressionLevelEnum(gzip.compression_level())),
       compression_strategy_(compressionStrategyEnum(gzip.compression_strategy())),
       memory_level_(memoryLevelUint(gzip.memory_level().value())),
-      window_bits_(windowBitsUint(gzip.window_bits().value())) {}
+      window_bits_(windowBitsUint(gzip.window_bits().value())),
+      chunk_size_(gzip.chunk_size().value() > 0 ? gzip.chunk_size().value() : 4096) {}
```
Comment (dio) [inline source/extensions/filters/http/gzip/gzip_filter.cc]: Great find! Let's have a ref/discussion/comment here how others set this value?

Comment (rgs1) [reply source/extensions/filters/http/gzip/gzip_filter.cc]: Sure -- let me add something

Comment (rgs1) [reply source/extensions/filters/http/gzip/gzip_filter.cc]: Hmm, maybe the proto file is a better place for this discussion/comment?

**Comment by dio** [comment]
Looks good. For the API review, @lizan could you help with this? Thanks!

File: source/extensions/filters/http/gzip/gzip_filter.cc (hunk 1 above)
Comment (dio) [reply source/extensions/filters/http/gzip/gzip_filter.cc]: 👍

File: api/envoy/config/filter/http/gzip/v2/gzip.proto [hunk 2]
```diff
@@ -95,4 +95,9 @@ message Gzip {
   // the fields `content_length`, `content_type`, `disable_on_etag_header` and
   // `remove_accept_encoding_header` are ignored.
   compressor.v2.Compressor compressor = 10;
+
+  // Value for Zlib's next output buffer. If not set, defaults to 4096.
+  // See https://www.zlib.net/manual.html for more details. Also see
+  // https://github.com/envoyproxy/envoy/issues/8448 for context on this filter's performance.
+  google.protobuf.UInt64Value chunk_size = 11 [(validate.rules).uint64 = {lte: 65536 gte: 4096}];
```
Comment (rojkov) [inline api/envoy/config/filter/http/gzip/v2/gzip.proto]: nit: I'd swap the `lte` and `gte` constraints to resemble a conventional interval.

Comment (rgs1) [reply api/envoy/config/filter/http/gzip/v2/gzip.proto]: everywhere else the convention seems to be `{lte: ... gte: ...}` (e.g.: `git grep `{gte` gives nothing with `lte` at the end). So let's stick to the convention.

Comment (rojkov) [reply api/envoy/config/filter/http/gzip/v2/gzip.proto]: Alright. nvm then

Comment (htuch) [inline api/envoy/config/filter/http/gzip/v2/gzip.proto]: Can I confirm the intention here is <= 2^16 rather than < 2^16?

Comment (htuch) [inline api/envoy/config/filter/http/gzip/v2/gzip.proto]: How come 64-bit value when the range is well within 32-bit?

Comment (rgs1) [reply api/envoy/config/filter/http/gzip/v2/gzip.proto]: I guess I went too far with future proofing -- I'll make it 32-bit. Thanks!

Comment (rgs1) [reply api/envoy/config/filter/http/gzip/v2/gzip.proto:102]: yeah <= 2^16

**Comment by htuch** [comment]
API LGTM, curious how this relates to https://github.com/envoyproxy/envoy/pull/10553? CC @rojkov

**Comment by rgs1** [comment]
> API LGTM, curious how this relates to #10553? CC @rojkov

We probably want to expose this there as well, but given it'll take a while for the gzip filter to go away we could start here.

**Comment by rojkov** [comment]
This PR looks good. I'm going to extend #10553 when this one is merged.

File: docs/root/intro/version_history.rst [hunk 3]
```diff
@@ -30,6 +30,9 @@ Version history
 * ext_authz: disabled the use of lowercase string matcher for headers matching in HTTP-based `ext_authz`.
   Can be reverted temporarily by setting runtime feature `envoy.reloadable_features.ext_authz_http_service_enable_case_sensitive_string_matcher` to false.
 * fault: added support for controlling abort faults with :ref:`HTTP header fault configuration <config_http_filters_fault_injection_http_header>` to the HTTP fault filter.
+* grpc-json: added support for building HTTP request into
```
Comment (lizan) [inline docs/root/intro/version_history.rst:33]: ?

Comment (rgs1) [reply docs/root/intro/version_history.rst:33]: hmm that probably happened during a merge (that changelog entry wasn't in the right place so I moved it around -- fixing now). Thanks for catching!

**Comment by rgs1** [comment]
is the coverage stuff still known to be flaky:

```
test/run_envoy_bazel_coverage.sh: line 43: 17958 Killed                  llvm-profdata merge -sparse -o ${COVERAGE_DATA} $(find -L bazel-out/k8-fastbuild/testlogs/test/coverage/coverage_tests/ -name coverage.dat)
```

?

**Comment by dio** [comment]
Seems like we need to merge master.

**Comment by rgs1** [comment]
Unrelated failure:

```
+ GCOV_PREFIX=/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/5160/execroot/envoy/_coverage/test/coverage/coverage_tests/shard_27_of_50/test
+ export LLVM_PROFILE_FILE=/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/5160/execroot/envoy/_coverage/test/coverage/coverage_tests/shard_27_of_50/test/%h-%p-%m.profraw
+ LLVM_PROFILE_FILE=/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/5160/execroot/envoy/_coverage/test/coverage/coverage_tests/shard_27_of_50/test/%h-%p-%m.profraw
+ [[ ! -z '' ]]
+ cd /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/5160/execroot/envoy/bazel-out/k8-fastbuild/bin/test/coverage/coverage_tests.runfiles/envoy
+ /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/5160/execroot/envoy/bazel-out/k8-fastbuild/bin/test/coverage/coverage_tests.runfiles/envoy/test/coverage/coverage_tests --gmock_default_mock_behavior=2 '--log-path /dev/null' '-l trace'
external/bazel_tools/tools/test/collect_coverage.sh: line 131: 30183 Segmentation fault      (core dumped) "$@"
+ TEST_STATUS=139
+ touch /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/sandbox/processwrapper-sandbox/5160/execroot/envoy/bazel-out/k8-fastbuild/testlogs/test/coverage/coverage_tests/shard_27_of_50/coverage.dat
+ [[ 139 -ne 0 ]]
+ echo --
```

File: api/envoy/config/filter/http/gzip/v2/gzip.proto [hunk 4]
```diff
@@ -95,4 +95,9 @@ message Gzip {
   // the fields `content_length`, `content_type`, `disable_on_etag_header` and
   // `remove_accept_encoding_header` are ignored.
   compressor.v2.Compressor compressor = 10;
+
+  // Value for Zlib's next output buffer. If not set, defaults to 4096.
+  // See https://www.zlib.net/manual.html for more details. Also see
+  // https://github.com/envoyproxy/envoy/issues/8448 for context on this filter's performance.
+  google.protobuf.UInt32Value chunk_size = 11 [(validate.rules).uint32 = {lte: 65536 gte: 4096}];
```
Comment (mattklein123) [inline api/envoy/config/filter/http/gzip/v2/gzip.proto:102]: We should make this change for v3 only.

File: source/extensions/filters/http/gzip/gzip_filter.cc (hunk 1 above)
Comment (mattklein123) [inline source/extensions/filters/http/gzip/gzip_filter.cc:32]: I think you want has_value() here. Better, use PROTOBUF_GET_WRAPPED_OR_DEFAULT

File: api/envoy/config/filter/http/gzip/v2/gzip.proto (hunk 4 above)
Comment (rgs1) [reply api/envoy/config/filter/http/gzip/v2/gzip.proto]: Done.

**Comment by rgs1** [comment]
@mattklein123 sorry about the force push :-( I made the mistake of fixing a merge conflict from the UI and forgot to pull back.

**Pull request merged**